gluon-private-wifi: add support for WPA3 encryption

blocktrron opened the pull request:
This commit adds WPA3 encryption for Gluon's Private WiFi feature.

It allows node operators to select between three encryption modes:
 - WPA2 (old standard - no MFP)
 - WPA2/WPA3 mixed-mode (optional MFP)
 - WPA3 (mandatory MFP)

MFP is supported by all drivers which are currently fully supported by Gluon (rt2800, mt76, ath9k, ath10k).

As we need to include a OpenSSL library for WPA3, this PR also includes a mechanic to define a featureset. This way, packages can be included at compile-time as well as the featureset retrieved at run-time to decide, which functions can be supported in a devices featureset.

This Pull-Request still needs documentation written, so It's labeled as RFC.

fixes #1826

blocktrron commented:
As a sidenote: We've seen issues connecting with a MFP enabled Access Point (WPA2/WPA3) on a 2012 Apple MacBook Pro. So we might want to also allow configuration of 802.11w instead of doing hard defaults here.

NeoRaider commented on scripts/target_lib.lua:
```diff
@@ -44,6 +45,7 @@ local default_options = {
 	aliases = {},
 	manifest_aliases = {},
 	packages = {},
+	featureset = false,
```
You could just define `featureset = "default"` here and undo all other changes in this file. Rather than adding a separate top-level function `featureset` for the defaults, this would be handled by the existing `defaults` function.

NeoRaider commented on package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua:
```diff
@@ -52,4 +52,8 @@ function M.is_outdoor_device()
 end
 
+function M.get_featureset()
+	return M.readfile("/lib/gluon/featureset")
```
The return value should be stripped to get rid of trailing whitespace.

NeoRaider commented on package/gluon-web-private-wifi/i18n/de.po:
```diff
@@ -16,6 +16,15 @@ msgstr "8-63 Zeichen"
+
+msgid "Required"
+msgstr "Erforderlich"
+
```
- `.pot` file should be updated using `contrib/i18n-scan.pl`
- All `.po` files should be updated using `msgmerge`

(see i18n docs)

NeoRaider commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,13 @@ Private WLAN
 It is possible to set up a private WLAN that bridges the WAN port and is separated from the mesh network.
 Please note that you should not enable ``mesh_on_wan`` simultaneously.
 
+The private WiFi is by default encrypted via WPA2. On devices with enough flash and a supported radio,
```
Weird word order, better: "The private Wifi is encrypted via WPA2 by default."

NeoRaider commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,13 @@ Private WLAN
+
+Additionally, IEEE 802.11w mangagement frame protection should be enforced for WPA2/WPA3, however this
+can lead to connectivity problems for older clients. In this case, mangagement frame protection can be
+set to optional or disabled completely using the config-mode. 
```
We usually refer to the web UI as "configuration wizard" or "advanced settings" instead of using the more technical term "config-mode" in the docs.

NeoRaider commented on docs/user/supported_devices.rst:
```diff
@@ -400,6 +400,10 @@ Footnotes
 
+.. [#featureset-tiny]
+  These devices do not support the full featureset of Gluon. Namely, WPA3 support
+  is missing on these devices.
```
Maybe we should mention that this is only relevant for the private-wifi feature at the moment.

blocktrron commented:
I've made the changes requested by @NeoRaider and also made the availability of WPA3 independent from the MFP support, as the support is now user-selectable anyhow.

rotanid commented:
maybe we should split up the featureset introduction in a separate PR?

blocktrron commented:
While this PR is mostly about the featuresets, i do not think that it is useful, as the usecase is directly connected with WPA3 support of any kind (OWE / SAE).

rotanid commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,13 @@ Private WLAN
+WPA3 or WPA2/WPA3 mixed-mode can be used instead of WPA2.
+
+Additionally, IEEE 802.11w mangagement frame protection should be enforced for WPA2/WPA3, however this
+can lead to connectivity problems for older clients. In this case, mangagement frame protection can be
```
two times: "mangagement" -> "management"

blocktrron replied:
Fixed.

rotanid commented on docs/user/supported_devices.rst:
```diff
@@ -400,6 +400,10 @@ Footnotes
 
+.. [#featureset-tiny]
+  These devices do not support the full featureset of Gluon. Namely, WPA3 support
+  for the Private-WiFi is missing on these devices.
```
why using the "WiFi" trademark here?

blocktrron replied:
Fixed

rotanid commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,13 @@ Private WLAN
 It is possible to set up a private WLAN that bridges the WAN port and is separated from the mesh network.
 Please note that you should not enable ``mesh_on_wan`` simultaneously.
 
+The private WiFi is encrypted via WPA2 by default. On devices with enough flash and a supported radio,
```
WiFi -> WLAN

blocktrron replied:
DOne

rotanid commented on package/gluon-web-private-wifi/i18n/de.po:
```diff
@@ -1,38 +1,45 @@
-"Hierfür wird der WAN-Port mit einem separaten WLAN gebridged. Diese "
-"Funktionalität ist völlig unabhängig von den Mesh-Funktionen des Knotens. "
-"Beachte, dass du nicht gleichzeitig das Meshen über den WAN-Port aktiviert "
-"haben solltest."
```
why is this section being removed?

blocktrron replied:
Because I'm to stupid to use tools correctly. Fixed now.

rotanid commented:
as WiFi is a trademark, i suggest to use WLAN instead where possible, but that's just my opinion

blocktrron commented:
@rotanid @NeoRaider Everything is (hopefully) addressed now.

rotanid commented on package/gluon-wireless-encryption/Makefile:
```diff
@@ -0,0 +1,33 @@
+
+define Package/gluon-wireless-encryption-wpa2
+  DEPENDS:=+hostapd-mini
+  TITLE:=Package for supporting WPA2 encrypted WiFi networks
```
WiFi -> WLAN

blocktrron replied:
fixed

rotanid commented on package/gluon-wireless-encryption/Makefile:
```diff
@@ -0,0 +1,33 @@
+
+define Package/gluon-wireless-encryption-wpa3
+  DEPENDS:=+hostapd-openssl
+  TITLE:=Package for supporting WPA3 encrypted WiFi networks
```
WiFi -> WLAN

blocktrron replied:
fixed

NeoRaider commented:
> it's not the scope of this PR, but we could think about renaming gluon-web-private-wifi to gluon-web-private-wlan

Or maybe *private-ap*?

rotanid commented:
> > it's not the scope of this PR, but we could think about renaming gluon-web-private-wifi to gluon-web-private-wlan
> 
> Or maybe _private-ap_?

also fine with me, but not wifi

blocktrron commented:
Fixed the two leftovers @rotanid pointed out and rebased.

The renaming of the `private-wifi` pacage is a good idea, but let's do this after this feature is finished ;)

NeoRaider commented on Makefile:
```diff
@@ -111,21 +111,30 @@ lint-lua: FORCE
+  $(1) := $(shell scripts/features.sh '$(2)' || echo '__ERROR__')
+endef
+$(eval $(call feature_packages, GLUON_FEATURESET_STANDARD_FEATURE_PACKAGES, $(GLUON_FEATURES) $(GLUON_FEATURESET_STANDARD_FEATURES)))
+$(eval $(call feature_packages, GLUON_FEATURESET_TINY_FEATURE_PACKAGES, $(GLUON_FEATURES) $(GLUON_FEATURESET_TINY_FEATURES)))
```
I think it would make sense to reuse `merge_packages` here (renamed to `merge_lists`), so the featureset featurelists can contain negative entries.

NeoRaider commented on .luacheckrc:
```diff
@@ -95,6 +95,7 @@ files["targets/*"] = {
 		"exec_capture_raw",
 		"exec_raw",
 		"factory_image",
+		"featureset",
```
I think this can be dropped?

NeoRaider commented on scripts/target_lib.lua:
```diff
@@ -44,6 +44,7 @@ local default_options = {
 	aliases = {},
 	manifest_aliases = {},
 	packages = {},
+	featureset = "standard",
```
Hmm, I wonder if "featureset" is really a good name for this - might get confusing, as we already have "features". And by itself, it categorizes devices rather than features - maybe something like "device class" (or just "class" in the target definitions) would be more appropriate?

NeoRaider commented on docs/user/site.rst:
```diff
@@ -527,8 +527,18 @@ GLUON_DEPRECATED
+    featureset.
+
+GLUON_FEATURESET_TINY_FEATURES
+    Defines a list of additional features to include for devices with the standard
```
"standard" -> "tiny"

NeoRaider commented on docs/user/supported_devices.rst:
```diff
@@ -414,6 +414,10 @@ Footnotes
 
+.. [#featureset-tiny]
+  These devices do not support the full featureset of Gluon. Namely, WPA3 support
+  for the private WLAN is missing on these devices.
```
Needs better explanation; should refer to the part of `site.rst` where setting the features for the featuresets is described.

NeoRaider replied:
We also need to document the criteria for setting the featuresets of devices somewhere.

As this PR is about adding WPA3 and the size of the OpenSSL libraries, I'd expect "tiny" to be about the flash size only, but somehow in the target definitions below you also set `tiny` for 8/32?

blocktrron commented:
@NeoRaider Hopefully I've catches everything. Also implemented device-class based site packages.

blocktrron commented:
rebased

NeoRaider commented on package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua:
```diff
@@ -52,4 +53,29 @@ function M.is_outdoor_device()
 end
 
+function M.supports_wpa3()
+	return unistd.access('/lib/gluon/features/wpa3')
```
Commit message needs to be updated.

NeoRaider replied:
Also, there is a typo in the title.

NeoRaider commented on package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua:
```diff
@@ -52,4 +53,29 @@ function M.is_outdoor_device()
+	uci:foreach('wireless', 'wifi-device', function()
+		local phypath = '/sys/kernel/debug/ieee80211/phy' .. idx .. '/'
+
+		if not util.file_contains_line(phypath .. 'hwflags', 'MFP_CAPABLE') then
```
Ugh, it sucks that there is no robust way to determine this without debugfs ...

mweinelt replied:
Besides checking the debug we can screenscrape `iw` output.

```
iw phy phy0 info | grep 00-0f-ac:6
```

via https://kevinlocke.name/bits/2019/12/28/checking-802.11w-support/
https://kevinlocke.name/bits/2019/12/28/checking-802.11w-support/

blocktrron replied:
I know the debugfs is not the must robust way, but i doubt screenscraping iw (which itself says you should not screenscrape it) is the better solution.

NeoRaider commented on package/gluon-core/luasrc/usr/lib/lua/gluon/util.lua:
```diff
@@ -35,6 +35,21 @@ function M.contains(table, value)
 end
 
+function M.file_contains_line(path, value)
+	local file_content = M.readfile(path)
```
- Use `io.lines()`: http://www.lua.org/manual/5.1/manual.html#pdf-io.lines
- Raise an error when the file can't be read (not sure if `io.lines()` already does that)

NeoRaider commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,14 @@ Private WLAN
 It is possible to set up a private WLAN that bridges the WAN port and is separated from the mesh network.
 Please note that you should not enable ``mesh_on_wan`` simultaneously.
 
+The private WLAN is encrypted through WPA2 by default. On devices with enough flash and a supported radio,
```
Better: "through" -> "using"

NeoRaider commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,14 @@ Private WLAN
+Additionally, IEEE 802.11w management frame protection should be enforced for WPA2/WPA3, however this
+can lead to connectivity problems for older clients. In this case, management frame protection can be
+made optional or completely disabled in the advanced settings tab.
+
```
Please add `:` after "docs" in the commit title.

NeoRaider commented on package/gluon-wireless-encryption/Makefile:
```diff
@@ -0,0 +1,33 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=gluon-wireless-encryption
+PKG_VERSION:=1
```
Drop `PKG_VERSION`, `PKG_RELEASE` is sufficient.

mweinelt replied:
`PKG_VERSION` is still here, did you miss this?

blocktrron replied:
Faulty push, fixed.

NeoRaider commented on Makefile:
```diff
@@ -111,7 +111,7 @@ lint-lua: FORCE
 	@scripts/lint-sh.sh
 
-GLUON_DEFAULT_PACKAGES := hostapd-mini
+GLUON_DEFAULT_PACKAGES :=
```
The variable `GLUON_DEFAULT_PACKAGES` and the `merge_lists` using it can be removed as well.

NeoRaider commented on package/gluon-wireless-encryption/Makefile:
```diff
@@ -0,0 +1,33 @@
+define Package/gluon-wireless-encryption-wpa2/install
+	$(INSTALL_DIR) $(1)/lib/gluon/features
+
+	touch $(1)/lib/gluon/features/wpa2
```
Having a `wpa2` feature at all does not make sense to me; making it a part of the `wpa3` as well package even less.

Let's drop the flag and the package, and directly reference `hostapd-mini` in `package/features`.

NeoRaider commented on Makefile:
```diff
@@ -115,19 +115,32 @@ GLUON_DEFAULT_PACKAGES := hostapd-mini
+
+$(eval $(call merge_lists, GLUON_DEFAULT_PACKAGES, $(GLUON_DEFAULT_PACKAGES) $(GLUON_SITE_PACKAGES)))
+$(eval $(call merge_lists, GLUON_CLASS_PACKAGES_standard, $(GLUON_CLASS_PACKAGES_standard) $(GLUON_FEATURE_PACKAGES_standard)))
+$(eval $(call merge_lists, GLUON_CLASS_PACKAGES_tiny, $(GLUON_CLASS_PACKAGES_tiny) $(GLUON_FEATURE_PACKAGES_tiny)))
```
- You can't use the same variable name as the target of `merge_lists` and as an input argument.
- Rather than initializing the different target variables as empty explicitly, this could be done in `merge_lists`.
- Remove all spaces around `,` in `$(call ...)` - probably not an issue in this case, but the spaces will be passed as parts of the arguments

NeoRaider commented on scripts/target_config_lib.lua:
```diff
@@ -7,6 +7,8 @@ return function(funcs)
 	local target = arg[1]
 	local extra_packages = arg[2]
+	local class_packages_standard = arg[3] or ''
+	local class_packages_tiny = arg[4] or ''
```
Make this a table:
```lua
local class_packages = {
  standard = arg[3],
  tiny = arg[4],
}
```
Then the code below can be simplified to
```lua
handle_pkgs(class_packages[dev.options.class])
```
by moving the ` or ''` and `string.gmatch()` into `handle_pkgs` as well.

NeoRaider replied:
It would be great it we could pass these lists in the environment instead of using separate arguments, but all ideas I have either limit future class names to characters that are valid in shell variable identifiers (disallowing `-` for example) or lead to more Makefile ugliness when generating the environment...

mweinelt replied:
I'm not sure how this was resolved either.

blocktrron replied:
Faulty push, fixed.

blocktrron commented:
Hopefully fixed everything. Will test my changes tomorrow and report back.

blocktrron commented:
Fixed two bugs with my changes, works as expected now.

Added the DIR-825 to the tiny device class, as I've missed it while researching the firmware partition sizes. Now all devices of the standard device class have at least 7M of flash space.

mweinelt commented on docs/dev/hardware.rst:
```diff
@@ -9,6 +9,18 @@ Having an ath9k, ath10k or mt76 based WLAN adapter is highly recommended,
+
+Device classes
+--------------
+Gluon is currently aware of two device classes. Depending on the device class, different
```
```suggestion
Gluon currently is aware of two device classes. Depending on the device class, different
```

mweinelt commented on docs/dev/hardware.rst:
```diff
@@ -9,6 +9,18 @@ Having an ath9k, ath10k or mt76 based WLAN adapter is highly recommended,
+Gluon is currently aware of two device classes. Depending on the device class, different
+features can be installed onto the device.
+
+The ``tiny`` device-class contains the following type of devices:
```
```suggestion
The ``tiny`` device-class contains devices with the following limitations:
```

mweinelt commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,14 @@ Private WLAN
 It is possible to set up a private WLAN that bridges the WAN port and is separated from the mesh network.
 Please note that you should not enable ``mesh_on_wan`` simultaneously.
 
+The private WLAN is encrypted via WPA2 by default. On devices with enough flash and a supported radio,
```
```suggestion
The private WLAN feature uses WPA2 encryption by default. On devices with enough flash and a supported radio,
```

blocktrron replied:
Altered this part according to @NeoRaider s previous feedback (which was overwritten by accident).

mweinelt commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,14 @@ Private WLAN
+WPA3 or WPA2/WPA3 mixed-mode can be used instead of WPA2. For this to work, the ``wireless-encryption-wpa3``
+feature has to be added to the list of your supported features.
+
+Additionally, IEEE 802.11w management frame protection should be enforced for WPA2/WPA3, however this
```
```suggestion
It is recommended to require IEEE 802.11w management frame protection for WPA2/WPA3 networks, however this
```

blocktrron replied:
done

mweinelt commented on docs/features/private-wlan.rst:
```diff
@@ -4,6 +4,14 @@ Private WLAN
+
+Additionally, IEEE 802.11w management frame protection should be enforced for WPA2/WPA3, however this
+can lead to connectivity problems for older clients. In this case, management frame protection can be
+set to optional or disabled completely using advanced settings. 
```
```suggestion
downgraded to be either optional or completely disabled using advanced settings. 
```

blocktrron replied:
Altered this part according to @NeoRaider s previous feedback (which was overwritten by accident).

mweinelt commented on docs/user/supported_devices.rst:
```diff
@@ -424,6 +424,12 @@ Footnotes
   in the next major release of Gluon is doubtful.
 
+.. [#device-class-tiny]
+  These devices only support a subset of Gluons capabilities due to size or memory
```
```suggestion
  These devices only support a subset of Gluons capabilities due to flash or memory size
```

blocktrron replied:
done

mweinelt commented on docs/user/supported_devices.rst:
```diff
@@ -424,6 +424,12 @@ Footnotes
 
+.. [#device-class-tiny]
+  These devices only support a subset of Gluons capabilities due to size or memory
+  constraints. Devices are classified as tiny in they provide less than 7M of useable
```
I'd prefer usable over useable.

blocktrron replied:
done

mweinelt commented on package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua:
```diff
@@ -55,4 +56,33 @@ function M.is_outdoor_device()
 end
 
+function M.device_supports_wpa3()
+	if M.match('ramips', 'rt305x') or M.match('ramips', 'mt7620') then
```
Is it possible to add a reference here why those exact targets are not supporting WPA3? How did you identify they don't?

blocktrron replied:
done

mweinelt commented on package/gluon-web-private-wifi/i18n/de.po:
```diff
@@ -13,18 +13,42 @@ msgstr ""
 msgstr "Privates WLAN"
 
+msgid "Required"
+msgstr "Aktiviert"
```
Erzwingen?

blocktrron replied:
Please not. MFP is a feature which can be *activated*, *deactivated* or flagged as *optional*. Erzwingen is not the correct term in this context.

mweinelt commented on package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua:
```diff
@@ -34,16 +54,24 @@ function f:write()
 				disabled   = false,
 			})
+
+			-- hostapd-mini won't start in case 802.11w is found in
```
```suggestion
			-- hostapd-mini won't start in case 802.11w is configured.
```

blocktrron replied:
done
